### PR TITLE
feat(docs): add consistent hash document

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title.

[Here is the preview](https://github.com/singularity-data/risingwave/blob/cyx/consistent-hash-doc/docs/consistent-hash.md).

### CODE REVIEW

File: docs/consistent-hash.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+### Streaming
+
+We know that a fragment has several actors as its different parallelisms, and that upstream actors will send data to downstream actors via [dispatcher](./streaming-overview.md#actors). The figure below illustrates how actors distribute data based on consistent hash by example. 
+
+![actor data distribution](./images/consistent-hash/actor-data.svg)
+
+In the figure, we can see that one upstream actor dispatches data to three downstream actors. The downstream actors are scheduled on the parallel units mentioned in previous example respectively. 
+
+Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta. It then decides how to send data by following steps:
```
Comment (fuyufjh) [inline docs/consistent-hash.md:51]:
```suggestion
Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:
```

File: docs/consistent-hash.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+When we perform parallel batch read, we should partition the data for each parallelism in a certain way. Now that we have vnodes corresponding to disjoint sets of data, this naturally forms a data partition pattern. That is to say, one vnode could be viewed as a minimal data partition unit, and we could aggregate several vnodes together to get a larger data partition.
+
+In vnode mapping, one parallel unit will correspond to several vnodes. Therefore, we could use parallel unit as the partition bound. Data falling to one parallel unit in vnode mapping will be one partition. This is better than range partition in that the partition is more stable when the primary key of a materialized view distributes non-randomly.
+
+### Storage
+
+If we look into the read-write pattern of streaming actors, we'll find that in most cases, actors only need to read the data written by itself (i.e. actor's internal states). That is to say, read data with the same vnodes as it previously writes. 
+
+Therefore, an instinctive way to place data in storage is to **group data by vnodes**. In this way, when actors perform read operation, it could touch as few SSTs as possible and thus trigger less I/O.
```
Comment (fuyufjh) [inline docs/consistent-hash.md:73]:
```suggestion
Therefore, an instinctive way to place data in storage is to **group data by vnodes**. In this way, when actors perform read operation, they could touch as few SST blocks as possible and thus trigger less I/O. 
```

**Review by fuyufjh — APPROVED** [review]
Nice work!!

**Review by hzxa21 — APPROVED** [review]
Thanks for the effort!

**Review by wcy-fdu — APPROVED** [review]
LGTM! The chart looks very clear!

**Review by Sunt-ing — APPROVED** [review]
Vivid illustrations. Accurate descriptions. Indeed, one of the best docs in our repo. Excellent!

File: docs/consistent-hash.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+### Streaming
+
+We know that a fragment has several actors as its different parallelisms, and that upstream actors will send data to downstream actors via [dispatcher](./streaming-overview.md#actors). The figure below illustrates how actors distribute data based on consistent hash by example. 
+
+![actor data distribution](./images/consistent-hash/actor-data.svg)
+
+In the figure, we can see that one upstream actor dispatches data to three downstream actors. The downstream actors are scheduled on the parallel units mentioned in previous example respectively. 
+
+Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:
```
Comment (BowenXiao1999) [inline docs/consistent-hash.md:51]: Is it possible that vnode mapping in CN is stale? How meta inform the latest to CN?

File: docs/consistent-hash.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+When scaling occurs, actors will be re-scheduled accordingly. By modifying the vnode mapping in meta and make streaming act on the new vnode mapping, we could minimize data movement from following aspects:
+
+- The data of existing actors will not be displaced too much. 
+- The block cache of a compute node will not be invalidated too much.
+
+### Batch
+
+When we perform parallel batch read, we should partition the data for each parallelism in a certain way. Now that we have vnodes corresponding to disjoint sets of data, this naturally forms a data partition pattern. That is to say, one vnode could be viewed as a minimal data partition unit, and we could aggregate several vnodes together to get a larger data partition.
+
+In vnode mapping, one parallel unit will correspond to several vnodes. Therefore, we could use parallel unit as the partition bound. Data falling to one parallel unit in vnode mapping will be one partition. This is better than range partition in that the partition is more stable when the primary key of a materialized view distributes non-randomly.
```
Comment (BowenXiao1999) [inline docs/consistent-hash.md:67]: > Therefore, we could use parallel unit as the partition bound.

What is use parallel unit as partition bound?

File: docs/consistent-hash.md (hunk 3 above)
Comment (xx01cyx) [reply docs/consistent-hash.md:51]: This is, in fact, not implemented yet. When scaling occurs, actors are designed to be dropped and recreated (currently), so the dispatchers would know the latest vnode mapping accordingly.

File: docs/consistent-hash.md (hunk 4 above)
Comment (xx01cyx) [reply docs/consistent-hash.md:67]: The vnodes that mapped to one parallel unit will form a partition group. Lemme give a clearer description about this in the doc.

**Review by TennyZhuang — APPROVED** [review]

**Comment by liurenjie1024** [comment]
cc @hengm3467 PTAL

**Review by xxchan — APPROVED** [review]

File: docs/consistent-hash.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+For all data $k \in U_k$, where $U_k$ is an unbounded set, we apply a hash function $v = H(k)$, where $v$ falls to a limited range. The hash function $H$ ensures that all data are hashed **uniformly** to that range. We call $v$ vnode, namely virtual node, as is shown as the squares in the figure below.
+
+![initial data distribution](./images/consistent-hash/data-distribution.svg)
+
+Then we have vnode mapping, which ensures that vnodes are mapped evenly to parallel units in the cluster. In other words, the number of vnodes that are mapped to each parallel unit should be as close as possible. This is denoted by different colors in the figure above. As is depicted, we have 3 parallel units (shown as circles), each taking $\frac{1}{3}$ of total vnodes. Vnode mapping is [constructed and maintained by meta](https://github.com/singularity-data/risingwave/blob/main/src/meta/src/manager/hash_mapping.rs).
+
+As long as the hash function $H$ could ensure uniformity, the data distribution determined by this strategy would be even across physical resources, even if data in $U_k$ skew to a certain range.
```
Comment (hengm3467) [inline docs/consistent-hash.md:29]: Minor rewording suggestion: "even if data in ... skew to a certain range" ----> "even if data in ... are skewed to a certain extent."

File: docs/consistent-hash.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+![actor state table](./images/consistent-hash/actor-state-table.svg)
+
+Now that we have 12 vnodes in total in the example, the data layout in storage will accordingly look like this:
+
+![storage data layout](./images/consistent-hash/storage-data-layout.svg)
+
+Note that we only show the logical sequence and aggregation of data in this illustration. The actual data may be separated into different SSTs in Hummock.
+
+Since the way that certain data are hashed to vnode is invariant, the encoding of the data will also be invariant. How we schedule the fragment (e.g. parallelism of the fragment) will not affect data encoding. In other words, storage will not care about vnode mapping, which is determined by meta and used only by streaming. This is actually a way of decoupling compute layer and storage layer.
```
Comment (hengm3467) [inline docs/consistent-hash.md:93]: Another minor rewording suggestion: the last sentence-----> "This is actually a way of decoupling the compute layer from the storage layer."

**Review by hengm3467 — APPROVED** [review]
Nice work! 
I just added a few very minor rewording suggestions. Otherwise looks good to me.

File: docs/consistent-hash.md (hunk 5 above)
Comment (xx01cyx) [reply docs/consistent-hash.md:29]: 1. Why use passive voice here? 🤔
2. "Range" here is sth like a line segment on a number axis. By "to a certain extent" it's more like "程度", which is not what I mean here.

Comment (hengm3467) [reply docs/consistent-hash.md:29]: Good question. When used as a verb, "skew" generally means causes something to be asymmetrical. In this sentence, data can only be the ones that are "caused" to appear in the way as-is. That's my understanding. I may be wrong. :) 

I misunderstood #2. Please ignore my comment.

Comment (xx01cyx) [reply docs/consistent-hash.md:29]: > data can only be the ones that are "caused" to appear in the way as-is.

I'll fix this. 

And I'll add an example to make clearer the meaning of "range".

**Pull request merged**